### PR TITLE
Feat/form3

### PR DESCRIPTION
**Pull request by emlindvall**
## Describe your changes:
[//]: <> (Which files were changed? What changes were made?)
- **Form.tsx**: Form component was created and rudimentarily styled. Functionality was added to show form container,  fetch jokes, hold error in state upon unsuccessful network request, allow user selection of jokes, hold a selected joke and selected image in state, save a new meme object, handle error upon click of "save meme" button if no joke has been selected, and for clearing joke input and hiding joke container upon click of "x" button.
- **App.tsx**: Changed to a class component in order to hold savedMemes array in state. 
- **Images.tsx**: Changed to class component in order to hold a selectedToggle boolean and an selectedImage path string in state, which allows conditional logic for opening and closing the form and populating the form with the user-selected image. 
- I**mage.tsx**: Altered to handle the onClick event that selects an image 
- SavedMemes.tsx: Class component created to hold saved meme array in state, accepting new saved memes from Form as props 
- **Meme.tsx**: Accepts the selected joke and selected image from Form to be rendered on  the page
- **Form.css**: Basic styling for the form component, when visible. 
- **Meme.css**: Basic styling for the user-created meme as it renders within the form 

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have tested the functionality of any new or updated features in the browser, if applicable
- [x] I have added a "Reviewer" to review & merge
- [x] If applicable, my code passes all tests

@niksseif @hfaerber
We've been fully integrating TypeScript into our application from the ground up! We're learning a ton about this new syntax, and about how to properly declare the type of each variable, component, prop, element, and event in order to create that additional cohesion within our code. We realize that this project probably isn't a realistic use-case for TypeScript, and that it would be much more helpful in a project whose scope was wider or more complex, but we are learning a lot as we go!

### CODE REVIEW

**Review by CodeMeister362 — APPROVED** [review]
Code is looking good! Im interested to go over the idea files.

**Review by sdtrachsel — APPROVED** [review]
Looks good! We may need to look into what command was used that updated the package.json and caused the errors to ensure there isn't other issues possibley floating around.

**Comment by mrlobatoman** [comment]
looks great

**Review by mrlobatoman — APPROVED** [review]
Looks great

**Comment by hfaerber** [comment]
Absolutely beautiful PR description of changes.  Thorough and helpful without being too much.  The description of changes was really helpful context to be able to effectively review the code changes.

File: src/apiCalls.tsx [hunk 1]
```diff
@@ -0,0 +1,15 @@
+const getJokes = () =>{
+  return fetch("https://api.api-ninjas.com/v1/dadjokes?limit=3", {
+    method: 'GET',
+    headers: {'X-Api-Key': 'C/1Bummp5MFkEZYvGxfhOQ==0XCoAUNJSOQoFpuv'}
+  })
+  .then((response) => {
+    if(!response.ok) {
+      throw new Error(`${response.status}`);
+    } else {
+      return response.json();
```
Comment (hfaerber) [inline src/apiCalls.tsx:10]: Nice response checking here!

File: src/components/form/Form.tsx [hunk 2]
```diff
@@ -1,12 +1,135 @@
...
+		});
+	}
+
+	getJokeOptions = () => {
+		getJokes()
+			.then((data) => {
+				const jokes: Joke[] = data.map((joke: Joke) => joke);
+				this.setState({ jokes: jokes });
+			})
+			.catch((error) => this.setState({ error: error.message }));
```
Comment (hfaerber) [inline src/components/form/Form.tsx:62]: Good use of .catch to update and track errors in state.  Good placement of the .catch here.

File: src/components/form/Form.tsx [hunk 3]
```diff
@@ -1,12 +1,135 @@
...
+	}
+
+	// lifecycle methods 
+	componentDidMount = () => {
+		this.getJokeOptions()
+	}
+
+	// methods
+	selectJoke = (joke: string) => {
+		this.setState({ selectedJoke: joke }, () => {
```
Comment (hfaerber) [inline src/components/form/Form.tsx:52]: Tell me about the 2nd argument you're passing here....

File: src/components/form/Form.tsx [hunk 4]
```diff
@@ -1,12 +1,135 @@
...
+	}
+
+	clearInputs = () => {
+		this.setState({ selectedImage: "" });
+		this.setState({ selectedJoke: "" });
+	}
+
+	// component render
+	render = () => {
+		const { jokes, selectedImage, selectedJoke } = this.state;
```
Comment (hfaerber) [inline src/components/form/Form.tsx:88]: great use of destructuring

File: src/components/form/Form.tsx [hunk 5]
```diff
@@ -1,12 +1,135 @@
...
+			<div className="generator-container">
+				<Meme
+					selectedJoke={selectedJoke}
+					selectedImage={selectedImage}
+				/>
+				<div className="form-wrapper">
+					<button className="close-button" onClick={this.props.closeForm}>X</button>
+					<form className="form-container">
+						<h4 className="joke-option-header">Choose Your Joke</h4>
+						{options}
```
Comment (hfaerber) [inline src/components/form/Form.tsx:122]: Love that you kept the map logic out of the return here

**Review by hfaerber — COMMENTED** [review]
Really nice start here.  Seeing lots of good things - love seeing the Typescript implementation!

**Pull request merged**